src/main/java/com/android/volley/toolbox/NetworkUtility.java

```diff
@@ -0,0 +1,224 @@
+        List<Header> combinedHeaders = HttpHeaderParser.combineHeaders(responseHeaders, entry);
+        return new NetworkResponse(
+                HttpURLConnection.HTTP_NOT_MODIFIED,
+                entry.data,
+                /* notModified= */ true,
+                SystemClock.elapsedRealtime() - requestStart,
+                combinedHeaders);
+    }
+
+    /** Reads the contents of an InputStream into a byte[]. */
+    static byte[] inputStreamToBytes(
+            @Nullable InputStream in, int contentLength, ByteArrayPool pool)
+            throws IOException, ServerError {
+        PoolingByteArrayOutputStream bytes = new PoolingByteArrayOutputStream(pool, contentLength);
+        byte[] buffer = null;
+        try {
+            if (in == null) {
+                throw new ServerError();
+            }
+            buffer = pool.getBuf(1024);
+            int count;
+            while ((count = in.read(buffer)) != -1) {
+                bytes.write(buffer, 0, count);
+            }
+            return bytes.toByteArray();
+        } finally {
+            try {
+                // Close the InputStream and release the resources by "consuming the content".
+                if (in != null) {
+                    in.close();
+                }
+            } catch (IOException e) {
+                // This can happen if there was an exception above that left the stream in
+                // an invalid state.
+                VolleyLog.v("Error occurred when closing InputStream");
+            }
+            pool.returnBuf(buffer);
+            bytes.close();
+        }
+    }
+
+    /**
+     * Attempts to prepare the request for a retry. If there are no more attempts remaining in the
+     * request's retry policy, a timeout exception is thrown.
+     *
+     * @param request The request to use.
+     */
+    private static void attemptRetryOnException(
+            final String logPrefix,
+            final Request<?> request,
+            @Nullable final AsyncNetwork.OnRequestComplete callback,
+            final VolleyError exception,
+            final Network network)
+            throws VolleyError {
+        final RetryPolicy retryPolicy = request.getRetryPolicy();
+        final int oldTimeout = request.getTimeoutMs();
+        try {
+            retryPolicy.retry(exception);
+        } catch (VolleyError e) {
+            request.addMarker(
+                    String.format("%s-timeout-giveup [timeout=%s]", logPrefix, oldTimeout));
+            throw e;
+        }
+        request.addMarker(String.format("%s-retry [timeout=%s]", logPrefix, oldTimeout));
+        if (network instanceof AsyncNetwork) {
+            if (callback == null) {
+                throw new VolleyError("No callback provided");
+            }
+            ((AsyncNetwork) network).performRequest(request, callback);
+        } else {
+            network.performRequest(request);
```

Comment:
I don't see this in the original BasicNetwork#attemptRetryOnException method. Why was it added here? (or maybe I'm missing where this is called in the original implementation)

Reply:
Yeah,  I don't think this quite works. BasicNetwork works today by running the request in a while(true) loop until this method throws; calling performRequest directly is sort of counter to that since it makes the loop redundant.

I think we either do the same thing in the async network (except we need to deal with asynchronicity) or do something like return a boolean here and let the caller inspect the result to determine if a retry is necessary, rather than passing the network in and making a call back out to it.

Reply:
In every case, either an error should be thrown (/passed through callback) or the request should be retried. Because of this, I just catch the error, call the callback, and return, or perform the request. Also, now BasicNetwork works the same as it originally did (either error is thrown to get out of while loop, or it's looped again).